File: 230.kth-smallest-element-in-a-bst.java
// @leet start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  int count = 0;
  int answer = 0;
  public int kthSmallest(TreeNode root, int k) {
    dfs(root, k);
    return answer;
  }
  public void dfs(TreeNode root, int k) {
    if (root == null) {
      return;
    }
    dfs(root.left, k);
    count++;
    if (count == k) {
      answer = root.val;
      return;
    }
    dfs(root.right, k);
  }
}
// @leet end

File: 230.kth-smallest-element-in-a-bst.py
# @leet start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k) -> int:
        count = [0]
        answer = [0]
        self.dfs(root, k, count, answer)
        return answer[0]

    def dfs(self, root, k, count, answer):
        if root is None:
            return
        self.dfs(root.left, k, count, answer)
        count[0] += 1
        if count[0] == k:
            answer[0] = root.val
            return
        self.dfs(root.right, k, count, answer)


# @leet end
